### ISSUE TITLE
Credo does not notice matches that are nested in conditions

### ISSUE THREAD

**Issue by TBK145** [rrrene/credo#77]
Credo will suggest to not use matches in eg. if-statements, however, if the match is nested credo will not find it.
Examples:
```
if !(x = Map.get(foo, :bar)), do: [...]
```
```
if (foo == (bar = :bar)), do: [...]
```

**Comment by rrrene** [comment]
Good catch!

Credo `v0.4.0` will allow the following:

      if value = values[field] do
        do_something_with_value
      end

while still disallowing more complex matches like these:

      if {:ok, %{"error" => true, "message" => message}} = parameter1 do
        do_error_handling
      end

So we could fix 

```
if (foo == ({:ok, bar} = some_function())), do: [...]
```

I am just not sure if the above is something people actually do (?)

**Comment by TBK145** [comment]
Ah, so will the deciding factor be the amount of assignments done, or the complexity?

**Comment by rrrene** [comment]
Everything except the "wildcard" assignemnt `x = foo(bar)` will cause an issue.

You can test this by pointing your Credo deps entry to `0.4.0-beta1`!

**Comment by TBK145** [comment]
So I checked the `release-0.4.0` branch, and I personally think this is a nice function.
I do think credo should check for nested matches, because there will be people who write
```
if !is_nil(baz = Map.get(foo, :bar)), do: baz
```

**Comment by rrrene** [comment]
Good point! :+1:

**Comment by rrrene** [comment]
Fixed in `v0.4.0-beta1`. Please report any further issues you might find :+1:
